Framework/Algorithms/src/RebinRagged2.cpp

```diff
@@ -0,0 +1,334 @@
+        outputWS->setHistogram(hist, builder.build());
+
+        prog.report();
+        PARALLEL_END_INTERRUPT_REGION
+      }
+      PARALLEL_CHECK_INTERRUPT_REGION
+
+      // Copy all the axes
+      for (int i = 1; i < inputWS->axes(); i++) {
+        outputWS->replaceAxis(i, std::unique_ptr<Axis>(inputWS->getAxis(i)->clone(outputWS.get())));
+        outputWS->getAxis(i)->unit() = inputWS->getAxis(i)->unit();
+      }
+
+      // Copy the units over too.
+      for (int i = 0; i < outputWS->axes(); ++i)
+        outputWS->getAxis(i)->unit() = inputWS->getAxis(i)->unit();
+      outputWS->setYUnit(eventInputWS->YUnit());
+      outputWS->setYUnitLabel(eventInputWS->YUnitLabel());
+    }
+  } // END ---- EventWorkspace
+
+  else
+
+  { //------- Workspace2D or other MatrixWorkspace ---------------------------
+    const bool isHist = inputWS->isHistogramData();
+
+    if (!isHist) {
+      // convert input to histogram
+      inputWS = inputWS->clone();
+      for (int hist = 0; hist < histnumber; ++hist) {
+        HistogramBuilder builder;
+        builder.setX(inputWS->histogram(hist).binEdges().rawData());
+        builder.setY(inputWS->readY(hist));
+        builder.setE(inputWS->readE(hist));
+        if (inputWS->hasDx(dist))
+          builder.setDx(inputWS->readDx(hist));
+        builder.setDistribution(dist);
+        inputWS->setHistogram(hist, builder.build());
+      }
+    }
+
+    // make output Workspace the same type as the input
+    outputWS = DataObjects::create<API::HistoWorkspace>(*inputWS, histnumber, inputWS->histogram(0));
+
+    // Copy over the 'vertical' axis
+    if (inputWS->axes() > 1)
+      outputWS->replaceAxis(1, std::unique_ptr<Axis>(inputWS->getAxis(1)->clone(outputWS.get())));
+
+    Progress prog(this, 0.0, 1.0, histnumber);
+
+    PARALLEL_FOR_IF(Kernel::threadSafe(*inputWS, *outputWS))
+    for (int hist = 0; hist < histnumber; ++hist) {
+      PARALLEL_START_INTERRUPT_REGION
+      auto xmin = xmins[hist];
+      auto xmax = xmaxs[hist];
+      const auto delta = deltas[hist];
+
+      HistogramData::BinEdges XValues_new(0);
+      static_cast<void>(VectorHelper::createAxisFromRebinParams({xmin, delta, xmax}, XValues_new.mutableRawData()));
+
+      outputWS->setHistogram(hist, HistogramData::rebin(inputWS->histogram(hist), XValues_new));
+      prog.report();
+      PARALLEL_END_INTERRUPT_REGION
+    }
+    PARALLEL_CHECK_INTERRUPT_REGION
+    outputWS->setDistribution(dist);
+
+    // Now propagate any masking correctly to the output workspace
+    // More efficient to have this in a separate loop because
+    // MatrixWorkspace::maskBins blocks multi-threading
+    for (int hist = 0; hist < histnumber; ++hist) {
+      if (inputWS->hasMaskedBins(hist)) // Does the current spectrum have any masked bins?
+      {
+        outputWS->setUnmaskedBins(hist);
+        this->propagateMasks(inputWS, outputWS, static_cast<int>(hist));
+      }
+    }
+    // Copy the units over too.
```

Comment:
Isn't this needed for the event case too?

Reply:
I was copying the logic from Rebin with how it copied the axes and units for different cases. But looking into this, none of this copying is need here or in Rebin because it is all handled by `DataObjects::create`. So I removed all the axes and units copying from Rebin and RebinRagged, we will see what happens.